src/Core/Managed/Shared/Extensibility/Implementation/Metrics/DependencyMetricsExtractor.cs

```diff
@@ -0,0 +1,361 @@
+                            //// In case of very high contention, this may happen a little more often,
+                            //// but will no longer happen once the MaxDependencyTypesToDiscover limit is reached.
+                            try
+                            {
+                                typeMetrics = thisMetrics.ByType.GetOrAdd(
+                                        dependencyType,
+                                        (depType) =>
+                                        {
+                                            lock (thisMetrics.TypeDiscoveryLock)
+                                            {
+                                                if (thisMetrics.DependencyTypesDiscoveredCount >= thisMetrics.MaxDependencyTypesToDiscover)
+                                                {
+                                                    throw new InvalidOperationException("MaxDependencyTypesToDiscover reached.");
+                                                }
+
+                                                thisMetrics.DependencyTypesDiscoveredCount++;
+
+                                                return new SucceessAndFailureMetrics(
+                                                    thisMetricManager.CreateMetric(
+                                                            MetricTerms.Autocollection.MetricNames.DependencyCall.Duration,
+                                                            new Dictionary<string, string>()
+                                                            {
+                                                                [MetricTerms.Autocollection.DependencyCall.PropertyNames.Success] = Boolean.TrueString,  // SUCCESS metric
+                                                                [MetricTerms.Autocollection.DependencyCall.PropertyNames.TypeName] = depType,
+                                                            }),
+                                                    thisMetricManager.CreateMetric(
+                                                            MetricTerms.Autocollection.MetricNames.DependencyCall.Duration,
+                                                            new Dictionary<string, string>()
+                                                            {
+                                                                [MetricTerms.Autocollection.DependencyCall.PropertyNames.Success] = Boolean.FalseString, // FAILURE metric
+                                                                [MetricTerms.Autocollection.DependencyCall.PropertyNames.TypeName] = depType,
+                                                            }));
+                                            }
+                                        });
+                            }
+                            catch (InvalidOperationException)
```

Comment:
I thought about this for a while. The problem is that CuncurrentDictionary.GetOrAdd does not allow the delegate to say "I changed my min, do not add anything". Null values are also allowed. For example, the following code:

            var cd = new ConcurrentDictionary<string, string>();

            cd.GetOrAdd("key1", (k) => k + "-value");
            cd.GetOrAdd("key2", (k) => null);

            Console.WriteLine("Entries in cd: " + cd.Count);

            try
            {
                cd.GetOrAdd("key3", (k) => throw new Exception("Not adding an element"));
            }
            catch (Exception ex)
            {
                Console.WriteLine("Observed error: " + ex);
            }

results in this output:

Entries in cd: 2
Observed error: System.Exception: Not adding an element
   at MiscTests01.Program.<>c.<Main>b__2_2(String k) in C:\00\Code\Sandbox\MiscTests01\MiscTests01\Program.cs:line 47
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at MiscTests01.Program.Main(String[] args) in C:\00\Code\Sandbox\MiscTests01\MiscTests01\Program.cs:line 47

So I cannot see a better way, but if you have a specific suggestion, I'd love to know. :)
In the meantime, let me expand on the exception message in case someone stumbles over the first chance... :)